RiotSwiftUI/Modules/Authentication/Registration/AuthenticationRegistrationModels.swift

```diff
@@ -0,0 +1,123 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+// MARK: View model
+
+enum AuthenticationRegistrationViewModelResult {
+    /// The user would like to select another server.
+    case selectServer
+    /// Validate the supplied username with the homeserver.
+    case validateUsername(String)
+    /// Create an account using the supplied credentials.
+    case createAccount(username: String, password: String)
+}
+
+// MARK: View
+
+struct AuthenticationRegistrationViewState: BindableState {
+    /// The address of the homeserver.
+    var homeserverAddress: String
+    /// Whether or not to show the username and password text fields with the next button
+    var showRegistrationForm: Bool
+    /// An array containing the available SSO options for login.
+    var ssoIdentityProviders: [SSOIdentityProvider]
+    /// View state that can be bound to from SwiftUI.
+    var bindings: AuthenticationRegistrationBindings
+    /// Whether or not the username field has been edited yet.
+    ///
+    /// This is used to delay showing an error state until the user has tried 1 username.
+    var hasEditedUsername = false
```

Comment:
Wouldnt this be true whenever `username.length > 0`, thus a computed property?

Reply:
The idea here was to show the textfield in an error state if the username is empty, but not when the screen is first shown.